### ISSUE TITLE
`flatten_parameters()` does not work with normed LSTM

### ISSUE THREAD

**Issue by erogol** [pytorch/pytorch#86457]
### 🐛 Describe the bug

flatten_parameters() func does not work as intended when an LSTM layer is passed to normalization functions like weight or spectral norms.  

Here is a sample script to reproduce the problem. 

```python
import torch
from torch import nn
from torch.nn import functional as F



class MyModule(nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.lstm = nn.LSTM(10, 20, 2)
        lstm_norm_fn_pntr = nn.utils.weight_norm
        self.lstm = lstm_norm_fn_pntr(self.lstm, "weight_hh_l0")

    def forward(self, x):
        self.lstm.flatten_parameters()
        self.lstm(x)
        return x


if __name__ == "__main__":
    m = MyModule()
    m.cuda()
    m.train()
    for _ in range(10):
        m(torch.randn(1, 10, 10).cuda())
```

```
/root/miniconda3/lib/python3.9/site-packages/torch/nn/modules/rnn.py:769: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters(). (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484809662/work/aten/src/ATen/native/cudnn/RNN.cpp:968.)
  result = _VF.lstm(input, hx, self._flat_weights, self.bias, self.num_layers,
```

### Versions

```
Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==1.12.1
[pip3] torch-tensorrt==1.2.0a0
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] pytorch                   1.12.1          py3.9_cuda11.3_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-tensorrt            1.2.0a0                  pypi_0    pypi
[conda] torchaudio                0.12.1               py39_cu113    pytorch
[conda] torchvision               0.13.1               py39_cu113    pytorch
```

**Comment by ngimel** [comment]
Duplicate of #39311, reparameterization recreates weights at evert step so they are no longer in the contiguous memory block, that's expected behavior.
